src/client/executor.rs

```diff
@@ -201,6 +226,18 @@ impl Client {
                 // release the selected server to decrement its operation count
                 drop(server);
 
+                if let Some(ref mut session) = session {
+                    if err.contains_label(TRANSIENT_TRANSACTION_ERROR) {
```

Comment:
I think this logic needs to live in `execute_operation_on_connection`, since otherwise retries that fail with these labels won't unpin on error.

Reply:
They'll probably need to be in both places, in that case - If the error causes the operation to unpin, it has to unpin _before_ the retry. But yeah, there's also the situation where the operation doesn't unpin, but the retry does.

Again, a good situation for a custom test case.

Reply:
Ah, right - because `execute_operation_on_connection` is called during the operation and on the retry... For some reason I was thinking that the method is only called on the first run.

That's my 5PM Thursday brain for ya...